Clustering Metrics Added

alchemist5617 opened the pull request:
I've implemented clustering metrics: 
mutual_info_score
normalized_mutual_info_score
adjusted_mutual_info_score
adjusted_rand_score
homogeneity_score
completeness_score
v_measure_score
I would really appreciate any feedback you can give me

Evizero commented:
Hello! Thanks for the interest in contributing and the effort you put into this.

I have to say I am unsure if this package is the best place for defining a contingency table. Have you seen https://github.com/nalimilan/FreqTables.jl ? Do you think we could build on the type defined there instead?

I think its worth delaying code-specific feedback/discussion until considering the above question. Some general feedback: I would rather avoid overly short function names; especially one or two letter ones.

alchemist5617 commented:
Hi, thanks for your feedback.

The problem with freqtable function of FreqTables.jl  is that, it doesn't work correctly when the input lists contain different types of data. Clustering metrics of Sklearn are able to work with these type of lists. I will raise this issue with them to see if they can fix it.    

In addition,I've fixed the error and changed the name of the functions.

Evizero commented:
Could you also see about the merge conflict? I changed the target branch to my rewrite branch of the package `refactor-0.6`. So if we are going to merge a version of your nice contribution it should be against that branch

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+and Correction for Chance, JMLR
+"""
+# ============================================================
+immutable ContingencyTable
```
going forward the keyword for this is `struct`

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+    n::Array{Int,2}
+    function ContingencyTable(U::AbstractVector,
+                              V::AbstractVector)
+    @_dimcheck length(U) == length(V)
```
for readability and to be somewhat in line to the official guideline, we should always indent the content of functions/loops etc

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+    number_of(item, list)
+number_of returns the number of the item in the list
+"""
+function number_of(item::Any,
```
if the there is no type for dispatch we usually leave it unspecified instead of `::Any`

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+"""
+function number_of(item::Any,
+                  list::AbstractVector)
+    length(find(list .== item))
```
Note that this allocates two temporary vectors.

1. the result of `list .== item`
2. the result of `find(...)`, which is only used to count how many elements it contains

You can write this without those allocations, by making use of the fact that vectors can be iterated over

```julia
count(elem -> elem == item, list)
```

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+"""
+    number_of_intersection(item_1, list_1, item_2, list_2)
+NumberOfIntersection returns the number of items in the intersection
+of first item in the first liat and second item in the second list.
```
typo: "first list"

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+                                list_2::AbstractVector)
+    logic_1 = item_1 .== list_1
+    logic_2 = item_2 .== list_2
+    length(find(logic_1 + logic_2 .== 2))
```
Similar to above this can be written more conservatively

```julia
count(elem -> elem == (item_1, item_2), zip(list_1, list_2))
```

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+
+"""
+    entropy(U)
+entropy(U) returns the entropy of clustering 'U'.
```
The official manual or Documenter (one of those anyway) suggest to phrase docstrings like this the following way:

```julia
"""
    entropy(U)

Return the entropy of clustering `U`
"""
```

Evizero replied:
(similar for the other docstrings)

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+entropy(U) returns the entropy of clustering 'U'.
+"""
+function entropy(U::AbstractVector)
+    result = 0
```
since below you add a `Float64` to `result`, it would be a lot better to set it to one here as well (for type stability) i.e.

```julia
result = 0.0
```

Note: If you haven't seen this before you can investigate type stability with

```
@code_warntype entropy(some_U)
```

where `some_U` is a vector you want to test the function with

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+    result = 0
+    N = length(U)
+    clusters = Set(U)
+    for c in clusters
```
if you don't need the variable `clusters` it is ok to just do `for c in Set(U)`. There is no overhead and its a little less noise

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+"""
+function entropy(U::AbstractVector,
+                 V::AbstractVector)
+    result = 0
```
same here: `result = 0.0`.

(and also the functions below)

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+    result
+end
+
+f(x) = factorial(x)
```
I'd rather not have this. I can see why you use it, but I don't think its worth it

Evizero commented on src/clustering.jl:
```diff
@@ -0,0 +1,351 @@
+                     sum(b .* ( b .- 1) ./ 2) / ( N * (N - 1) / 2)
+    max_index = 0.5 * (sum(a .* (a .- 1) ./ 2) + sum(b .* (b .- 1) ./ 2))
+    if (max_index - expected_index) == 0
+        return 0
```
This should also be `0.0` to be the same type as the other branch of the `if`
